sphinx/builders/__init__.py

```diff
@@ -464,8 +504,20 @@ def read_doc(self, docname: str) -> None:
         if path.isfile(docutilsconf):
             self.env.note_dependency(docutilsconf)
 
+        filename = self.env.doc2path(docname)
+        filetype = get_filetype(self.app.config.source_suffix, filename)
+        publisher = self._get_publisher(filetype)
+
         with sphinx_domains(self.env), rst.default_role(docname, self.config.default_role):
-            doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
+            # set up error_handler for the target document
+            error_handler = UnicodeDecodeErrorHandler(docname)
+            codecs.register_error('sphinx', error_handler)  # type: ignore
+
+            publisher.set_source(source_path=filename)
+            publisher.publish()
```

Comment:
I'm afraid reusing docutils components between multiple documents. At the first glance, the document object is generated at every build. So it's not shared. But reader, parser and writer objects are shared. Do you know these objects do not store any status? It seems nobody resets the status of them on the `publish()` call.

Reply:
The reader, parser and writer objects do store state, but it is overwritten on every call to `.publish()`. 

We could explicitly reset the state of the objects after each call?

A

Reply:
Thank you for your explanation. Okay, let's go with this.